org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -745,6 +760,51 @@ protected IStatus run(IProgressMonitor monitor) {
 		return CompletableFuture.completedFuture(this.languageServer);
 	}
 
+	/**
+	 * Dispatches a LS request on a single executor thread tied to this server. Use of this guarantees that the server
+	 * will have seen all previous such requests (and document updates), and will not receive any further updates
+	 * prior to receiving this one.
+	 * @param <T> LS response type
+	 * @param fn LS method to invoke
+	 * @return Async result
+	 */
+	@NonNull
+	<T> CompletableFuture<T> executeOnLatestVersion(@NonNull Function<LanguageServer, ? extends CompletionStage<T>> fn) {
```

Comment:
Do we want something else than "onLatestVersion"? If not, we may just omit the suffix and call it `requestLanguageServer` and `notifiyLanguageServer`.
The javadoc shouldn't really mention things about how stuff is dispatch, it should be a bit higher level. Technical comments can be added as `@implNote` to the Javadoc IIRC.

Reply:
I agree, the names are confusing, especially if we are making the concurrency guarantees intrinsic to the core API rather than just adding in extra methods with optional extra guarantees. I think I favour `execute(fn)` and `sendNotification(fn)` - short and simple; `notify` is a bit problematic to use on its own as a method name because of `Object.notify()`!

Maybe the technical notes about the threading design could be a markdown document at the top level?